[KEYCLOAK-9045] Add CORS headers on proxied upstream responses

fredbi opened the pull request:
* fixes #9045
~~* adds CorsDebug option~~
* adds CorsDisableUpstream option

Signed-off-by: Frederic BIDON <frederic@oneconcern.com>

fredbi commented:
Got some difficulties with gofmt on config.go with go11. Some gofmt fix will be needed at some point here.

fredbi commented on doc.go:
```diff
@@ -278,7 +278,8 @@ type Config struct {
 	CorsMaxAge time.Duration `json:"cors-max-age" yaml:"cors-max-age" usage:"max age applied to cors headers (Access-Control-Max-Age)"`
-
+	// CorsDebug enables debug messages in the CORS middleware
+	CorsDebug bool `json:"cors-debug" yaml:"cors-debug" usage:"enable cors debugi logging"`
```
ok, this sounds better to have a single flag. Will reuse Verbose.

fredbi commented on server.go:
```diff
@@ -603,6 +604,7 @@ func (r *oauthProxy) createUpstreamProxy(upstream *url.URL) error {
 
 	// create the forwarding proxy
 	proxy := goproxy.NewProxyHttpServer()
+	proxy.KeepDestinationHeaders = true
```
Why not, but what is the use case for value false? In the current setup, headers setup by the middleware are just thrown away in the relayed response from upstream (cors or not cors btw).  I am not convinced of any proper usage with a "false" setting. Further, setting up CORS, and defaulting to making it unusable for upstream looks at best difficult to explain to a regular user.

fredbi replied:
I am changing this to making it settable as you suggested, but I'd rather change the default, which would otherwise be misleading IHMO.
Like this: 
```golang
proxy.KeepDestinationHeaders = !cfg.CorsDisableUpstream
```

fredbi commented on e2e_test.go:
```diff
@@ -0,0 +1,171 @@
+	if err != nil {
+		return err
+	}
+	//t.Log("INFO:starting gatekeeper")
```
ok done

fredbi commented on e2e_test.go:
```diff
@@ -0,0 +1,171 @@
+				"jwks_uri":"http://`+e2eOauthListener+`/auth/realms/master/protocol/openid-connect/certs"
+			}`)
+		}
+		//t.Logf("INFO: handling: %s", e2eOauthURL)
```
ok done

fredbi commented on server.go:
```diff
@@ -184,6 +184,7 @@ func (r *oauthProxy) createReverseProxy() error {
 			AllowCredentials: r.config.CorsCredentials,
 			ExposedHeaders:   r.config.CorsExposedHeaders,
 			MaxAge:           int(r.config.CorsMaxAge.Seconds()),
+			Debug:            r.config.CorsDebug,
```
sure. done

fredbi commented on e2e_test.go:
```diff
@@ -0,0 +1,165 @@
+	_, err := checkListen.Get(endpoint)
+	limit := 0
+	for err != nil && limit < maxWaitCycles {
+		//fmt.Printf("INFO:waiting for %s to start... (%d)\n", endpoint, limit)
```
I still have to remove this one as well...

abstractj replied:
@fredbi please, go for it.

fredbi replied:
done

abstractj commented:
Hi @fredbi, I'm putting these changes on hold because I'm taking some time off starting today and these changes require more testing. Thank you for the time spent on this, I'm looking forward for more contributions next year.

Merry Christmas to you and your family!

fredbi commented:
@abstractj sure! Enjoy your Christmas!
More PR are in preparation...  so we'll have some dealings next year!

fredbi commented:
@abstractj  I factored in the remark from #456 and reverted on the optionality of setting this behavior to upstream (I admit I am focused on the reverse-proxy use case, not the forward proxy).
I also added a fix to override possible cors headers returned from upstream (gatekeeper headers rule).

I let this as separate commits atm, and I am ready to squash them if you want.

abstractj commented:
@fredbi could you please squash your commits into a single one?

fredbi commented:
@abstractj  done, thank you

Aside: you certainly know that github repo owners can now automatically squash PRs when merging

EronWright commented:
@abstractj gentle reminder to take another look at this.

aflag commented:
We've been having issues with this as well. It would be nice if it could be merged and released.

javefang commented:
Could we look to get this one merged? 

(@fredbi one minor thing I found while trying to build this, it looks like master branch moved on and switched from using "dep" to "go modules", might worth a rebase?)

fredbi commented:
@javefang master branch has barely moved since my latest rebase on this one, which does not cause any conflict.

@aflag most of my contributions here are pretty old and have never been merged or seriously discussed. It just looks like this repo is slowly dying for some unknown reason. I am currently actively maintaining a fork here:  https://github.com/oneconcern/keycloak-gatekeeper

abstractj commented:
Hi @fredbi my apologies if the pull-requests are not being merged in a timely manager. Unfortunatelly we've been busy with not only Gatekeeper, but also with other projects. We take contributions very serious, that's the reason why sometimes it take that long. Just doing a blind merge of changes from contributions, without any testing or reviewing would be disrespectful.

Thanks for all the changes you sent. Some them were already merged and acknowledged here: https://github.com/keycloak/keycloak-documentation/blob/master/release_notes/topics/7_0_0.adoc

aflag commented:
@abstractj I appreciate that, but this is an important fix (note that, without this, the session cookie becomes nearly useless, as it's never updated after it's refreshed) and it seems to have the tests passing. Also, it seems like the changes requested were made. I don't think it would be disrespectful or careless to merge them at this point. Quite the opposite.

sharebear commented:
I'm unable to log in to the jira so commenting here... the jira issue is marked with fix version 7.0.0 but it looks to me like this was merged three days after 7.0.0 was tagged and therefore this fix isn't in a released version yet.

javefang commented:
agreed, it is not in the 7.0.1 release either. @abstractj is there any plan to make a release to include this fix? without the fix, every api call after the initial access_token expiry will trigger the gatekeeper to do a token refersh... and we are forced to build our own gatekeeper from source at the moment...

abstractj commented:
> I'm unable to log in to the jira so commenting here... the jira issue is marked with fix version 7.0.0 but it looks to me like this was merged three days after 7.0.0 was tagged and therefore this fix isn't in a released version yet.

For questions, please ask on the user mailing list. Things can be easily missed if we discuss things on merged PRs.
